File: usaco_submissions/Section 2/2.4/fractions_to_decimal.py
"""
TASK: fracdec
USER: Vignesh Iyer
LANG: PYTHON3
"""
from decimal import *
getcontext().prec = 100

def read_input(filename):
  global n
  global d
  with open(filename, 'r') as fin:
      n,d = tuple(map(int, fin.readline().strip().split()))

def get_first_occurrence(c, i, d):
  for j in range(i, len(d)):
    if(d[j] == c):
      return j
  return -1

def get_repeated_seq(d):
  l = len(d)
  for i in range(l):
    if(d[i] == '.'):
      break
  i += 1
  
  for j in range(i, l):
    start = j
    end = get_first_occurrence(d[j], start+1, d)
    if(end == -1):
      continue
    found = True
    cnt = 0
    for k in range(start, end):
      if(end + cnt < l and d[k] != d[end + cnt]):
        found = False  
        break
      cnt += 1
    if(found):
      ans = d[:i] + d[i:start] 
      if(d[start:end] == '0'):
        return ans
      return ans + '(' + d[start:end] + ')'
  return d

def solve(fout):
  if(n%d == 0):
    ans = str(n/d)
  else:
    ans = str(Decimal(n/d))
    ans = get_repeated_seq(ans)
  
  for i in range(len(ans)):
    fout.write(ans[i])
    if((i > 0 and i%75 == 0) or i == len(ans)-1):
      fout.write('\n')

filename = 'fracdec'
read_input(filename+'.in')

## output
with open(filename+'.out', 'w') as fout:
  solve(fout)

File: usaco_submissions/Section 2/2.4/fractions_to_decimal.cpp
#include <bits/stdc++.h>
using namespace std;

/*
TASK: fracdec
USER: Vignesh Iyer
LANG: C++
*/

#define ll long long
#define nline "\n"
#define F first
#define S second
typedef pair<int, int> pii;

void read_input(string filename){
    freopen((filename + ".in").c_str(), "r", stdin);
    freopen((filename + ".out").c_str(), "w", stdout);
}

int N, D;

void solve(){
    cin >> N >> D;
    if(N%D == 0) cout << N/D << ".0" << nline;
    else{
        cout << N/D << ".";
        int char_count = 2;
        char_count += to_string(N/D).length();
        unordered_map<int, int> hm;
        vector<int> order;
        unordered_set<int> seen;
        N %= D;
        int end = -1;
        while(true){
            N *= 10;
            if(seen.count(N)){
                end = N;
                break;
            }
            seen.insert(N);
            hm[N] = N/D;
            order.push_back(N);
            if(N%D == 0){
                for(int n : order){
                    if(char_count%76 == 1) cout << nline;
                    cout << hm[n];
                    char_count++;
                }
                cout << nline;
                return;
            }
            N %= D;
        }
        for(int n : order){
            if(char_count%76 == 1) cout << nline;
            if(n == end){
                cout << "(";
                char_count++;
            }
            cout << hm[n];
            char_count++;
        }
        cout << ")\n";
    }
}

int main() {
    read_input("fracdec");
    int tc = 1;
    // cin >> tc;
    while(tc--)
        solve();

    return 0;
}